### PR TITLE
SpectralSolverRZ.H: Move Implementations to Source File

### PR DESCRIPTION
**Pull request by RemiLehe**
**These code changes were extracted from #1216. Many thanks to @dpgrote for implementing them.**

This PR simply moves routines from the `.H` to the corresponding `.cpp` ; the actual code is unchanged.

### CODE REVIEW

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H [hunk 1]
```diff
@@ -117,24 +96,8 @@ class SpectralSolverRZ
...
-         * derived classes, from objects of class SpectralSolverRZ through the private
-         * unique pointer \c algorithm.
-         *
-         * \param[in,out] current Array of unique pointers to \c MultiFab storing
-         *                        the three components of the current density
-         */
-        void VayDeposition (std::array<std::unique_ptr<amrex::MultiFab>,3>& current)
-        {
-            algorithm->VayDeposition(field_data, current);
-        }
```
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H]: @RemiLehe Thank you again for refactoring #1216. I think this interface for the function `VayDeposition` was removed by accident here.

Comment (RemiLehe) [reply Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H]: Oops! Thanks a lot for pointing this out! I just updated the PR.

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp [hunk 2]
```diff
@@ -50,3 +52,97 @@ SpectralSolverRZ::SpectralSolverRZ(amrex::BoxArray const & realspace_ba,
                                      n_rz_azimuthal_modes, lev);
 
 };
+
+
+/* \brief Transform the component `i_comp` of MultiFab `field_mf`
+ *  to spectral space, and store the corresponding result internally
+ *  (in the spectral field specified by `field_index`) */
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp]: They only belong in the header file:
```suggestion
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp [hunk 3]
```diff
@@ -50,3 +52,97 @@ SpectralSolverRZ::SpectralSolverRZ(amrex::BoxArray const & realspace_ba,
...
+SpectralSolverRZ::ForwardTransform (amrex::MultiFab const & field_mf,
+                                    int const field_index,
+                                    int const i_comp) {
+    WARPX_PROFILE("SpectralSolverRZ::ForwardTransform");
+    field_data.ForwardTransform(field_mf, field_index, i_comp);
+};
+
+/* \brief Transform the two MultiFabs `field_mf1` and `field_mf2`
+ *  to spectral space, and store the corresponding results internally
+ *  (in the spectral field specified by `field_index1` and `field_index2`) */
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp]:
```suggestion
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp [hunk 4]
```diff
@@ -50,3 +52,97 @@ SpectralSolverRZ::SpectralSolverRZ(amrex::BoxArray const & realspace_ba,
...
+void
+SpectralSolverRZ::ForwardTransform (amrex::MultiFab const & field_mf1, int const field_index1,
+                                    amrex::MultiFab const & field_mf2, int const field_index2) {
+    WARPX_PROFILE("SpectralSolverRZ::ForwardTransform");
+    field_data.ForwardTransform(field_mf1, field_index1,
+                                field_mf2, field_index2);
+};
+
+/* \brief Transform spectral field specified by `field_index` back to
+ * real space, and store it in the component `i_comp` of `field_mf` */
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp]:
```suggestion
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp [hunk 5]
```diff
@@ -50,3 +52,97 @@ SpectralSolverRZ::SpectralSolverRZ(amrex::BoxArray const & realspace_ba,
...
+void
+SpectralSolverRZ::BackwardTransform (amrex::MultiFab& field_mf,
+                                     int const field_index,
+                                     int const i_comp) {
+    WARPX_PROFILE("SpectralSolverRZ::BackwardTransform");
+    field_data.BackwardTransform(field_mf, field_index, i_comp);
+};
+
+/* \brief Transform spectral fields specified by `field_index1` and `field_index2`
+ * back to real space, and store it in `field_mf1` and `field_mf2`*/
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp]:
```suggestion
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp [hunk 6]
```diff
@@ -50,3 +52,97 @@ SpectralSolverRZ::SpectralSolverRZ(amrex::BoxArray const & realspace_ba,
...
+void
+SpectralSolverRZ::BackwardTransform (amrex::MultiFab& field_mf1, int const field_index1,
+                                     amrex::MultiFab& field_mf2, int const field_index2) {
+    WARPX_PROFILE("SpectralSolverRZ::BackwardTransform");
+    field_data.BackwardTransform(field_mf1, field_index1,
+                                 field_mf2, field_index2);
+};
+
+/* \brief Update the fields in spectral space, over one timestep */
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp]:
```suggestion
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp [hunk 7]
```diff
@@ -50,3 +52,97 @@ SpectralSolverRZ::SpectralSolverRZ(amrex::BoxArray const & realspace_ba,
...
+    // Virtual function: the actual function used here depends
+    // on the sub-class of `SpectralBaseAlgorithm` that was
+    // initialized in the constructor of `SpectralSolverRZ`
+    algorithm->pushSpectralFields(field_data);
+};
+
+/**
+  * \brief Public interface to call the member function ComputeSpectralDivE
+  * of the base class SpectralBaseAlgorithmRZ from objects of class SpectralSolverRZ
+  */
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp]:
```suggestion
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp [hunk 8]
```diff
@@ -50,3 +52,97 @@ SpectralSolverRZ::SpectralSolverRZ(amrex::BoxArray const & realspace_ba,
...
+/**
+ * \brief Public interface to call the virtual function \c CurrentCorrection,
+ * defined in the base class SpectralBaseAlgorithmRZ and possibly overridden
+ * by its derived classes (e.g. PsatdAlgorithmRZ), from
+ * objects of class SpectralSolverRZ through the private unique pointer \c algorithm
+ *
+ * \param[in,out] current two-dimensional array of unique pointers to MultiFab
+ *                        storing the three components of the current density
+ * \param[in]     rho     unique pointer to MultiFab storing the charge density
+ */
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp]:
```suggestion
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp [hunk 9]
```diff
@@ -50,3 +52,97 @@ SpectralSolverRZ::SpectralSolverRZ(amrex::BoxArray const & realspace_ba,
...
+
+/**
+ * \brief Public interface to call the virtual function \c VayDeposition,
+ * declared in the base class SpectralBaseAlgorithmRZ and defined in its
+ * derived classes, from objects of class SpectralSolverRZ through the private
+ * unique pointer \c algorithm.
+ *
+ * \param[in,out] current Array of unique pointers to \c MultiFab storing
+ *                        the three components of the current density
+ */
```
Comment (ax3l) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp]:
```suggestion
```

**Review by ax3l — APPROVED** [review]
Thanks! :sparkles:

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H [hunk 10]
```diff
@@ -36,49 +36,30 @@ class SpectralSolverRZ
...
-        void ForwardTransform (amrex::MultiFab const & field_mf,
-                               int const field_index,
-                               int const i_comp=0 ) {
-            BL_PROFILE("SpectralSolverRZ::ForwardTransform");
-            field_data.ForwardTransform(field_mf, field_index, i_comp);
-        }
+
+        void ForwardTransform(amrex::MultiFab const & field_mf,
+                              int const field_index,
+                              int const i_comp=0);
```
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H]:
```suggestion
        void ForwardTransform (amrex::MultiFab const & field_mf,
                               int const field_index,
                               int const i_comp=0);
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H [hunk 11]
```diff
@@ -36,49 +36,30 @@ class SpectralSolverRZ
...
          *  to spectral space, and store the corresponding results internally
          *  (in the spectral field specified by `field_index1` and `field_index2`) */
-        void ForwardTransform (amrex::MultiFab const & field_mf1, int const field_index1,
-                               amrex::MultiFab const & field_mf2, int const field_index2) {
-            BL_PROFILE("SpectralSolverRZ::ForwardTransform");
-            field_data.ForwardTransform(field_mf1, field_index1,
-                                        field_mf2, field_index2);
-        }
+        void ForwardTransform(amrex::MultiFab const & field_mf1, int const field_index1,
+                              amrex::MultiFab const & field_mf2, int const field_index2);
```
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H]:
```suggestion
        void ForwardTransform (amrex::MultiFab const & field_mf1, int const field_index1,
                               amrex::MultiFab const & field_mf2, int const field_index2);
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H [hunk 12]
```diff
@@ -36,49 +36,30 @@ class SpectralSolverRZ
...
          * real space, and store it in the component `i_comp` of `field_mf` */
-        void BackwardTransform (amrex::MultiFab& field_mf,
-                                int const field_index,
-                                int const i_comp=0) {
-            BL_PROFILE("SpectralSolverRZ::BackwardTransform");
-            field_data.BackwardTransform(field_mf, field_index, i_comp);
-        }
+        void BackwardTransform(amrex::MultiFab& field_mf,
+                               int const field_index,
+                               int const i_comp=0);
```
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H]:
```suggestion
        void BackwardTransform (amrex::MultiFab& field_mf,
                                int const field_index,
                                int const i_comp=0);
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H [hunk 13]
```diff
@@ -36,49 +36,30 @@ class SpectralSolverRZ
...
         /* \brief Transform spectral fields specified by `field_index1` and `field_index2`
          * back to real space, and store it in `field_mf1` and `field_mf2`*/
-        void BackwardTransform (amrex::MultiFab& field_mf1, int const field_index1,
-                                amrex::MultiFab& field_mf2, int const field_index2) {
-            BL_PROFILE("SpectralSolverRZ::BackwardTransform");
-            field_data.BackwardTransform(field_mf1, field_index1,
-                                         field_mf2, field_index2);
-        }
+        void BackwardTransform(amrex::MultiFab& field_mf1, int const field_index1,
+                               amrex::MultiFab& field_mf2, int const field_index2);
```
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H]:
```suggestion
        void BackwardTransform (amrex::MultiFab& field_mf1, int const field_index1,
                                amrex::MultiFab& field_mf2, int const field_index2);
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H [hunk 14]
```diff
@@ -36,49 +36,30 @@ class SpectralSolverRZ
...
 
         /* \brief Update the fields in spectral space, over one timestep */
-        void pushSpectralFields () {
-            BL_PROFILE("SpectralSolverRZ::pushSpectralFields");
-            // Virtual function: the actual function used here depends
-            // on the sub-class of `SpectralBaseAlgorithm` that was
-            // initialized in the constructor of `SpectralSolverRZ`
-            algorithm->pushSpectralFields(field_data);
-        }
+        void pushSpectralFields();
```
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H]:
```suggestion
        void pushSpectralFields ();
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H [hunk 15]
```diff
@@ -102,10 +83,8 @@ class SpectralSolverRZ
           * \brief Public interface to call the member function ComputeSpectralDivE
           * of the base class SpectralBaseAlgorithmRZ from objects of class SpectralSolverRZ
           */
-        void ComputeSpectralDivE ( const std::array<std::unique_ptr<amrex::MultiFab>,3>& Efield,
-                                   amrex::MultiFab& divE ) {
-            algorithm->ComputeSpectralDivE( field_data, Efield, divE );
-        }
+        void ComputeSpectralDivE(const std::array<std::unique_ptr<amrex::MultiFab>,3>& Efield,
+                                 amrex::MultiFab& divE);
```
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H]:
```suggestion
        void ComputeSpectralDivE (const std::array<std::unique_ptr<amrex::MultiFab>,3>& Efield,
                                  amrex::MultiFab& divE);
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H [hunk 16]
```diff
@@ -117,10 +96,8 @@ class SpectralSolverRZ
          *                        storing the three components of the current density
          * \param[in]     rho     unique pointer to MultiFab storing the charge density
          */
-        void CurrentCorrection ( std::array<std::unique_ptr<amrex::MultiFab>,3>& current,
-                                 const std::unique_ptr<amrex::MultiFab>& rho ) {
-             algorithm->CurrentCorrection( field_data, current, rho );
-        }
+        void CurrentCorrection(std::array<std::unique_ptr<amrex::MultiFab>,3>& current,
+                               const std::unique_ptr<amrex::MultiFab>& rho);
```
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.H]:
```suggestion
        void CurrentCorrection (std::array<std::unique_ptr<amrex::MultiFab>,3>& current,
                                const std::unique_ptr<amrex::MultiFab>& rho);
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp [hunk 17]
```diff
@@ -50,3 +52,63 @@ SpectralSolverRZ::SpectralSolverRZ(amrex::BoxArray const & realspace_ba,
...
+    // on the sub-class of `SpectralBaseAlgorithm` that was
+    // initialized in the constructor of `SpectralSolverRZ`
+    algorithm->pushSpectralFields(field_data);
+};
+
+void
+SpectralSolverRZ::ComputeSpectralDivE (const std::array<std::unique_ptr<amrex::MultiFab>,3>& Efield,
+                                       amrex::MultiFab& divE) {
+    algorithm->ComputeSpectralDivE(field_data, Efield, divE);
+};
```
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp]:
```suggestion
}
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp [hunk 18]
```diff
@@ -50,3 +52,63 @@ SpectralSolverRZ::SpectralSolverRZ(amrex::BoxArray const & realspace_ba,
...
+
+void
+SpectralSolverRZ::pushSpectralFields () {
+    WARPX_PROFILE("SpectralSolverRZ::pushSpectralFields");
+    // Virtual function: the actual function used here depends
+    // on the sub-class of `SpectralBaseAlgorithm` that was
+    // initialized in the constructor of `SpectralSolverRZ`
+    algorithm->pushSpectralFields(field_data);
+};
```
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp]:
```suggestion
}
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp [hunk 19]
```diff
@@ -50,3 +52,63 @@ SpectralSolverRZ::SpectralSolverRZ(amrex::BoxArray const & realspace_ba,
...
+    field_data.BackwardTransform(field_mf, field_index, i_comp);
+};
+
+void
+SpectralSolverRZ::BackwardTransform (amrex::MultiFab& field_mf1, int const field_index1,
+                                     amrex::MultiFab& field_mf2, int const field_index2) {
+    WARPX_PROFILE("SpectralSolverRZ::BackwardTransform");
+    field_data.BackwardTransform(field_mf1, field_index1,
+                                 field_mf2, field_index2);
+};
```
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp]:
```suggestion
}
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp [hunk 20]
```diff
@@ -50,3 +52,63 @@ SpectralSolverRZ::SpectralSolverRZ(amrex::BoxArray const & realspace_ba,
...
+                                field_mf2, field_index2);
+};
+
+void
+SpectralSolverRZ::BackwardTransform (amrex::MultiFab& field_mf,
+                                     int const field_index,
+                                     int const i_comp) {
+    WARPX_PROFILE("SpectralSolverRZ::BackwardTransform");
+    field_data.BackwardTransform(field_mf, field_index, i_comp);
+};
```
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp]:
```suggestion
}
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp [hunk 21]
```diff
@@ -50,3 +52,63 @@ SpectralSolverRZ::SpectralSolverRZ(amrex::BoxArray const & realspace_ba,
...
+    field_data.ForwardTransform(field_mf, field_index, i_comp);
+};
+
+void
+SpectralSolverRZ::ForwardTransform (amrex::MultiFab const & field_mf1, int const field_index1,
+                                    amrex::MultiFab const & field_mf2, int const field_index2) {
+    WARPX_PROFILE("SpectralSolverRZ::ForwardTransform");
+    field_data.ForwardTransform(field_mf1, field_index1,
+                                field_mf2, field_index2);
+};
```
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp]:
```suggestion
}
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp [hunk 22]
```diff
@@ -50,3 +52,63 @@ SpectralSolverRZ::SpectralSolverRZ(amrex::BoxArray const & realspace_ba,
...
 };
+
+
+void
+SpectralSolverRZ::ForwardTransform (amrex::MultiFab const & field_mf,
+                                    int const field_index,
+                                    int const i_comp) {
+    WARPX_PROFILE("SpectralSolverRZ::ForwardTransform");
+    field_data.ForwardTransform(field_mf, field_index, i_comp);
+};
```
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp]:
```suggestion
}
```

File: Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp [hunk 23]
```diff
@@ -50,3 +52,63 @@ SpectralSolverRZ::SpectralSolverRZ(amrex::BoxArray const & realspace_ba,
...
+SpectralSolverRZ::ComputeSpectralDivE (const std::array<std::unique_ptr<amrex::MultiFab>,3>& Efield,
+                                       amrex::MultiFab& divE) {
+    algorithm->ComputeSpectralDivE(field_data, Efield, divE);
+};
+
+void
+SpectralSolverRZ::CurrentCorrection (std::array<std::unique_ptr<amrex::MultiFab>,3>& current,
+                                      const std::unique_ptr<amrex::MultiFab>& rho) {
+     algorithm->CurrentCorrection(field_data, current, rho);
+};
```
Comment (EZoni) [inline Source/FieldSolver/SpectralSolver/SpectralSolverRZ.cpp]:
```suggestion
}
```

**Review by EZoni — CHANGES_REQUESTED** [review]
@RemiLehe Thanks again for this PR! I made only two sets of suggestions to crosscheck:
1. in SpectralSolverRZ.H the white space between function names and brackets `(` was deleted by accident and I think it is useful to keep it for `git grep` purposes;
2. in SpectralSolverRZ.cpp I think the `;` after the function implementations are not necessary (I believe they would raise the warnings `extra ';' [-Wpedantic]`).

**Review by EZoni — APPROVED** [review]

**Pull request merged**